examples/apps/tree-comparison/src/model/legacyTreeInventoryList.ts

```diff
@@ -307,3 +322,48 @@ export const LegacyTreeInventoryListFactory = new DataObjectFactory<LegacyTreeIn
 	[LegacySharedTree.getFactory()],
 	{},
 );
+
+export const LegacyTreeInventoryListFactoryNew = new DataObjectFactory<NewTreeInventoryList>(
+	"legacy-tree-inventory-list",
+	NewTreeInventoryList,
+	[LegacySharedTree.getFactory(), SharedTree.getFactory()],
+	{},
+	undefined,
+	undefined,
+	async (runtime, root) => {
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		const tree = await root
+			.get<IFluidHandle<LegacySharedTree>>(legacySharedTreeKey)! // ! it's actually "LegacySharedTree | SharedTree"
+			.get();
+
+		// ! Future TODO: If we loaded legacy shared tree factory, we can assume we'll need converter code
+		if (tree.attributes.type === SharedTreeFactory.Type) {
+			const rootNode = tree.currentView.getViewNode(tree.currentView.root);
+			const inventoryItemsNodeIds = tree.currentView
+				// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+				.getViewNode(rootNode.traits.get("inventory" as TraitLabel)![0])
+				.traits.get("inventoryItems" as TraitLabel);
+
+			const inventoryItems: LegacyTreeInventoryItem[] = [];
+			if (inventoryItemsNodeIds !== undefined) {
+				for (const inventoryItemNodeId of inventoryItemsNodeIds) {
+					const inventoryItemNode = tree.currentView.getViewNode(inventoryItemNodeId);
+					inventoryItems.push(makeInventoryItemFromInventoryItemNode(tree, inventoryItemNode));
+				}
+			}
+			const newSharedTree = SharedTree.create(runtime);
+			const view = newSharedTree.viewWith(treeConfiguration);
+			view.initialize(
+				new InventorySchema({
+					inventoryItemList: inventoryItems.map((val) => ({
+						id: val.id,
+						name: val.name,
+						quantity: val.quantity,
+					})),
+				}),
+			);
+			view.dispose();
+			root.set(legacySharedTreeKey, newSharedTree.handle);
```

Comment:
The prototype does not cover / answer what happens if two clients do the same conversion at the same time - they would step on each other and overwrite that key, leaving one of the clients working with an instance of the tree that is not reachable anymore.

And as noted in other comments, we also should have good story for case where client loaded with old data model and observes conversion op - the client can't do much (I think) - it would be really hard for users of the system to reload data model and redo all the bindings, so I think it's fine if client simply crashes when such state happens, as long as it can reload and not get back to exactly same state.